File: presentation/src/main/java/bg/government/virusafe/app/utils/validators/EgnValidator.java
package bg.government.virusafe.app.utils.validators;

import org.threeten.bp.LocalDate;
import org.threeten.bp.temporal.ChronoUnit;

import java.util.ArrayList;
import java.util.Arrays;
import java.util.List;

import androidx.annotation.Nullable;

import static com.upnetix.applicationservice.registration.model.Gender.VALUE_FEMALE;
import static com.upnetix.applicationservice.registration.model.Gender.VALUE_MALE;

/**
 * Validate EGN.
 */
public class EgnValidator implements PersonalIdValidator {

    private static final List<Integer> EGN_WEIGHTS = Arrays.asList(2, 4, 8, 5, 10, 9, 7, 3, 6);
    private static final Integer EGN_MOD = 11;

    private String personalNumber;
    private boolean isValid = false;
    private Long years = null;
    private String gender;

    public boolean isValidPersonalId() {

        return isValid;
    }

    @Nullable
    public Long getYears() {

        return years;
    }

    @Nullable
    public String getGender() {

        return gender;
    }

    public void initPersonalNumber(@Nullable String personalNumber) {

        this.personalNumber = personalNumber;
        isValid = isValidPersonalNumber();
        if (!isValid) {
            years = null;
            gender = null;
        }
    }

    public void clearAll() {

        this.personalNumber = null;
        isValid = false;
        years = null;
        gender = null;
    }

    private boolean isValidPersonalNumber() {

        if (personalNumber == null || personalNumber.equals("")) {
            return true;
        } else if (personalNumber.length() != 10) {
            return false;
        }

        final char[] charArr = personalNumber.toCharArray();
        final List<Integer> personalNumberDigits = new ArrayList<>(charArr.length);

        for (char c : charArr) {
            personalNumberDigits.add(c - '0');
        }

        updateGender(personalNumberDigits);

        boolean isValidBirthday = validateBirthday(personalNumberDigits);
        boolean isValidEgnChecksum = validateEgnCheckSum(personalNumberDigits);
        return isValidBirthday && isValidEgnChecksum;
    }

    private void updateGender(List<Integer> egnDigits) {

        int c = egnDigits.get(8);
        if (c % 2 == 0) {
            gender = VALUE_MALE;
        } else {
            gender = VALUE_FEMALE;
        }
    }

    private boolean validateBirthday(List<Integer> egnDigits) {

        int year = egnDigits.get(0) * 10 + egnDigits.get(1);
        int month = egnDigits.get(2) * 10 + egnDigits.get(3);
        int day = egnDigits.get(4) * 10 + egnDigits.get(5);

        // Handle month offsets based on birth years before 1900 and after 2000.
        if (month > 40) {
            month -= 40;
            year += 2000;
        } else if (month > 20) {
            month -= 20;
            year += 1800;
        } else {
            year += 1900;
        }
        try {
            LocalDate birthDate = LocalDate.of(year, month, day);
            LocalDate now = LocalDate.now();
            years = ChronoUnit.YEARS.between(birthDate, now);
        } catch (Exception e) {
            return false;
        }
        return true;
    }

    private boolean validateEgnCheckSum(List<Integer> egnDigits) {

        int checkSum = 0;
        for (int i = 0; i < egnDigits.size() - 1; i++) {
            checkSum += (egnDigits.get(i) * EGN_WEIGHTS.get(i));
        }
        checkSum %= EGN_MOD;
        if (checkSum == 10) {
            checkSum = 0;
        }
        return egnDigits.get(9) == checkSum;
    }
}

File: presentation/src/main/java/bg/government/virusafe/app/utils/validators/EgnValidator.kt
@file:Suppress("MagicNumber")
package bg.government.virusafe.app.utils.validators

import bg.government.virusafe.app.utils.empty
import com.upnetix.applicationservice.registration.model.Gender
import org.threeten.bp.LocalDate
import org.threeten.bp.temporal.ChronoUnit
import java.util.*

/**
 * Validate EGN.
 */
private const val EGN_MOD = 11

class EgnValidator : PersonalIdValidator {

	private val egnWeights = listOf(2, 4, 8, 5, 10, 9, 7, 3, 6)
	private var personalNumber: String? = null
	private var isValid = false

	override var years: Long? = null
		private set

	override var gender: String? = null
		private set

	override fun isValidPersonalId(): Boolean = isValid

	override fun initPersonalNumber(personalNumber: String?) {
		this.personalNumber = personalNumber
		isValid = isValidPersonalNumber()
		if (!isValid) {
			years = null
			gender = null
		}
	}

	override fun clearAll() {
		personalNumber = null
		isValid = false
		years = null
		gender = null
	}

	@Suppress("ReturnCount")
	private fun isValidPersonalNumber(): Boolean {
		if (personalNumber == null || personalNumber == String.empty) {
			return true
		} else if (personalNumber!!.length != 10) {
			return false
		}

		val charArr = personalNumber!!.toCharArray()
		val personalNumberDigits: MutableList<Int> = ArrayList(charArr.size)

		for (c in charArr) {
			personalNumberDigits.add(c - '0')
		}

		updateGender(personalNumberDigits)

		val isValidBirthday: Boolean = validateBirthday(personalNumberDigits)
		val isValidEgnChecksum: Boolean = validateEgnCheckSum(personalNumberDigits)

		return isValidBirthday && isValidEgnChecksum
	}

	private fun updateGender(egnDigits: List<Int>) {
		val c = egnDigits[8]
		gender = if (c % 2 == 0) {
			Gender.VALUE_MALE
		} else {
			Gender.VALUE_FEMALE
		}
	}

	@Suppress("TooGenericExceptionCaught, LongMethod")
	private fun validateBirthday(egnDigits: List<Int>): Boolean {
		var year = egnDigits[0] * 10 + egnDigits[1]
		var month = egnDigits[2] * 10 + egnDigits[3]
		val day = egnDigits[4] * 10 + egnDigits[5]

		// Handle month offsets based on birth years before 1900 and after 2000.
		when {
			month > 40 -> {
				month -= 40
				year += 2000
			}
			month > 20 -> {
				month -= 20
				year += 1800
			}
			else -> {
				year += 1900
			}
		}

		years = try {
			val birthDate = LocalDate.of(year, month, day)
			val now = LocalDate.now()
			ChronoUnit.YEARS.between(birthDate, now)
		} catch (e: Exception) {
			return false
		}

		return true
	}

	private fun validateEgnCheckSum(egnDigits: List<Int>): Boolean {
		var checkSum = 0
		for (i in 0 until egnDigits.size - 1) {
			checkSum += egnDigits[i] * egnWeights[i]
		}

		checkSum %= EGN_MOD
		if (checkSum == 10) {
			checkSum = 0
		}

		return egnDigits[9] == checkSum
	}
}
